app/common/components/Pagination/index.js

```diff
@@ -0,0 +1,204 @@
+            url={this.getPageUrl(page)}
+            current={page === currentPage}
+          />
+        ))}
+        <PageDirection
+          type="next"
+          page={currentPage + 1}
+          url={this.getPageUrl(currentPage + 1)}
+          hide={currentPage === totalPages}
+        />
+      </ul>
+    );
+  }
+
+  get leftPages() {
+    const { marginPages, totalPages } = this.props;
+
+    const length = Math.min(marginPages, totalPages);
+
+    return valuesFromRange(length);
+  }
+
+  get middlePages() {
+    const { pageRange, marginPages, currentPage, totalPages } = this.props;
+
+    if (this.shouldRangeBeTruncated) {
+      if (this.isCurrentPageInLeftRange) {
+        // 1 2 | 3 4 5 (6) _ | 14 15
+        return valuesFromRange(pageRange - TRUNCATE_LENGTH, marginPages);
+      } else if (this.isCurrentPageInRightRange) {
+        // 1 2 _ (10) 11 12 13 | 14 15
+        return valuesFromRange(
+          pageRange - TRUNCATE_LENGTH,
+          this.rightRange.begin
+        );
+      } else {
+        // 1 2 | _ 7 (8) 9 _ | 14 15
+        const length = pageRange - 2 * TRUNCATE_LENGTH;
+
+        return valuesFromRange(length, currentPage - Math.ceil(length / 2));
+      }
+    } else {
+      // 1 (2) || 3
+      // 1 (2) || 3 4
+      // 1 (2) | 3 | 4 5
+      // 1 (2) | 3 4 5 6 7 | 8 9
+
+      const length = Math.max(totalPages - marginPages * 2, 0);
+
+      return valuesFromRange(length, marginPages);
+    }
+  }
+
+  get rightPages() {
+    const { totalPages, marginPages } = this.props;
+
+    const restPages = Math.max(totalPages - marginPages, 0);
+    const length = Math.min(marginPages, restPages);
+
+    return valuesFromRange(length, restPages);
+  }
+
+  get shouldRangeBeTruncated() {
+    const { marginPages, pageRange, totalPages } = this.props;
+
+    const maxPagesLength = marginPages * 2 + pageRange;
+
+    return maxPagesLength < totalPages;
+  }
+
+  get isCurrentPageInLeftRange() {
+    return (
+      this.state.currentPage <= this.leftRange.end - this.props.rangeOverlap
+    );
+  }
+
+  get isCurrentPageInRightRange() {
+    return (
+      this.state.currentPage > this.rightRange.begin + this.props.rangeOverlap
+    );
+  }
+
+  get leftRange() {
+    const { marginPages, pageRange } = this.props;
+
+    return {
+      begin: 0,
+      end: marginPages + pageRange - TRUNCATE_LENGTH
+    };
+  }
+
+  get rightRange() {
+    const { totalPages, marginPages, pageRange } = this.props;
+
+    return {
+      begin: totalPages - (marginPages + pageRange - TRUNCATE_LENGTH),
+      end: totalPages
+    };
+  }
+
+  getPageUrl(page) {
+    const { location: { pathname, query } } = this.props;
+
+    return createUrl(pathname, { ...query, page });
+  }
+
+  handleClick = event => {
+    event.preventDefault();
+
+    const { page } = event.target.dataset;
+
+    if (page) {
+      this.setState({ currentPage: parseInt(page, 10) });
+    }
+  };
+}
+
+const valuesFromRange = (length, offset = 0) =>
```

Comment:
Maybe we should move this to separate array utils/helpers?

Reply:
It makes sense. Let's do this when similar functionality will be required in other components